Dynamic keymap implementation

sebhz opened the pull request:
This patch allows specifying the full key mapping from an external resource file.
Maps for US keyboard as well as French keyboard (MS Windows and Linux) are provided.

No configuration added for the joystick (yet)

sebhz commented:
Overall this patchset it pretty early - submitting it more for comments than for actual merge (if you want to merge it, fine, though !). I tested it under MS Windows with a Windows FR map and lo and behold all my keys are now working !

ColinPitrat commented on src/keyboard.cpp:
```diff
@@ -891,308 +722,489 @@ int kbd_layout[4][KBD_MAX_ENTRIES][2] = {
+
+// Format of a line: CPC_xxx\tSDLK_xxx\tMODIFIER
+// Last field is optional
+void parse_line(char *s, unsigned int line)
```
It would be nice to have some unit tests for this method and the next one

ColinPitrat commented:
Build is failing on tests with:
test/CapriceVKeyboard.cpp:62:20: error: ambiguous overload for 'operator=' (operand types are 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}' and 'int')
   CPC.kbd_layout = 0;
                    ^

Overall this looks fine (I'll have a deeper look).
For merging we shouldn't break features though (vkeyboard & configuring from GUI)

ColinPitrat commented on src/gui/src/CapriceVKeyboard.cpp:
```diff
@@ -119,21 +119,22 @@ namespace wGui {
         escaped = false;
         cap32_cmd = false;
       } else {
+// TODO Fix this !
```
I have the habit of putting TODO(cpitrat) (so you could use TODO(sebhz)).
Not mandatory but this may help you find your TODOs in the code more easily.

sebhz commented on cap32.cfg:
```diff
@@ -142,11 +142,10 @@ pp_device=0
+#   Host keyboard layout map
+#   Name of the map file (under the resources directory)
+#   If the file is not used a US keyboard is assumed
+kbd_layout=keymap_fr_linux.map
```
Yes it does. After all CPC is a UK machine :-)

sebhz commented on src/gui/src/CapriceOptions.cpp:
```diff
@@ -247,6 +247,7 @@ CapriceOptions::CapriceOptions(const CRect& WindowRect, CWindow* pParent, CFontE
     m_pDropDownCPCLanguage->SelectItem(CPC.keyboard);
     m_pDropDownCPCLanguage->SetIsFocusable(true);
     // option 'kbd_layout' which is the platform keyboard layout (i.e. the PC keyboard layout)
+    /*
```
Done - we build the dropdown list by listing all the .map files.
The is still the possibility of a race condition if a map file is added to the resources directory after the  GUI is displayed - but I do not really know how to solve this simply.

ColinPitrat replied:
Adding a file is not an issue (it will not be available in options, that's all).
Removing one could be though (well, not even a big one, it will default to UK I think). But if someone mess-up with resources while running the program, he is looking for problems isn't he ?

sebhz commented:
Overall this is working as before, except keymap is now configurable. I must confess I did not do an extensive testing of the whole french maps.
There is still some legacy cleanup to do and some nice additions to make:

- we should probably use maps all over the board (not sure if the overhead is big or not - are maps implemented as hashtables in C++ ?) rather than have maps and tables.
- remapping joysticks would also be nice... though I think this comes basically for free right now, as we can simply define JOY_x_UP, DOWN etc... in the map file. However it would collide with joystick emulation. I still think joystick emulation is a nicer way to handle the joystick -> you remap your keys when you need it only.
- some more tests to add as mentioned above
- A spanish keymap would also be a nice addition.

ColinPitrat commented:
Thanks, I will make my best to review it thoroughly and give it a try in the coming days.

sebhz commented:
Sure, thanks. Do not bother looking at all the commits individually - surprisingly enough the changes are not that big if you look at all the files.

sebhz commented:
Plans for this:

- Address all the maintainer comment and have this merged :-)
- Once merged, replace table lookups by C++ unordered_map lookups, which will make the code more readable, simpler (no need to maintain 4 tables for each modifiers), and extend functionalities by allowing more modifiers (NUMLOCK to remap joysticks for example)
- Finally, convert keyboard.cpp into a keymapper class, with a few simple interfaces, and hide all the gory remapping there. All we need are methods to get SDL_Keysym from character (to get characters from the virtual keyboard) and get CPC key from SDL_Keysym (to input characters into the emulator). No need to expose all those tables and maps.

ColinPitrat commented:
To get it to work under linux, I had to `dos2unix` the files.
I think this should work under Windows with files in Unix format, can you check ?
If we allow to generate the file, it's something to keep in mind though.

ColinPitrat commented on src/keyboard.cpp:
```diff
@@ -891,308 +724,520 @@ int kbd_layout[4][KBD_MAX_ENTRIES][2] = {
+	std::filebuf fb;
+	char line[MAX_LINE_LENGTH]; // sufficient for now ! TODO (sebhz): proper malloc'ing etc...
+
+	if (layout_file.empty() || (fb.open(layout_file, std::ios::in) == nullptr)) {
```
Do you need to check for emptyness ? Open would fail with empty string wouldn't it ?

sebhz replied:
Right. Also, the way the layout file name is constructed, it will default to something non empty anyway.

ColinPitrat commented on src/keyboard.cpp:
```diff
@@ -891,308 +724,520 @@ int kbd_layout[4][KBD_MAX_ENTRIES][2] = {
+		return true;
+}
+
+// TODO (sebhz) there must be a way to to this more efficiently !
```
If you mean performances, I don't think we care at this place (not a critical section)

sebhz replied:
I simply meant that there is probably a way to do a simple pointer assignment for this, rather than iterating over the table to to a copy of a constnt table, but C++ seems to be pretty strict about this. Anyway, if we end up using maps all over the board this will eventually disappear. So I removed the comment also.

ColinPitrat commented on src/keyboard.cpp:
```diff
@@ -456,10 +478,112 @@ dword cpc_kbd[3][149] = {
-    { CPC_0,          SDLK_0 },
+std::map<char, CPC_KEYS> CPCkeysFromChars = {
+    // Char to CPC keyboard translation
+	// TODO(sebhz): How on Earth do we map non ASCII chars ?
```
Which non-ASCII chars are you thinking about ?

ColinPitrat commented on src/fileutils.cpp:
```diff
@@ -57,3 +57,19 @@ std::vector<std::string> listDirectory(std::string sDirectory) {
+
+// Returns a vector containing the names of the files having extension "ext" in
+// the specified directory
+std::vector<std::string> listDirectoryExt(std::string sDirectory, std::string ext) {
```
Pass sDirectory and ext as `const std::string&`.
We could rename sDirectory just directory here and in listDirectory.

sebhz replied:
I cannot make sDirectory const, as it is modified in the listDirectory function. Other than this, I did the modifications.

ColinPitrat commented on src/gui/src/CapriceOptions.cpp:
```diff
@@ -247,15 +249,22 @@ CapriceOptions::CapriceOptions(const CRect& WindowRect, CWindow* pParent, CFontE
-    m_pDropDownPCLanguage->SetIsFocusable(true);
 
+    mapFileList = listDirectoryExt(CPC.resources_path, "map");
+	unsigned int currentMapIndex = 0;
```
Indentation is weird here until the end of the for (mix of tabs & spaces ?)

sebhz replied:
Right. I fixed this to use the old style (no tab). BTW this code has a small problem: if the layout file is wrong/empt/undefined, the first layout file found will be selected by default, which is a bit misleading. I do not really know how to handle this.

ColinPitrat commented:
Another remark, when I use `keyboard_fr_win.map` and I type "azerty" on my keyboard it displays "qzerty" where I'd expect "qwerty". Both keys 'z' and 'w' display 'z'. When pressing shift this is not the case.

And yet the file seems correct ... Adding a few logs of keysyms received and mapping done I couldn't find the reason for that.

sebhz commented:
UNIX end of lines do not seem to create any problem under MS Windows. I addressed all your comments (I think !), including this one. Regarding the problem with "qzerty" where "qwerty" is expected, I'll have a look at it later on this week (no linux machine handy these days). The only explanation I can hazard is that SDL keymap under windows is more broken than I thought, but it is hardly convincing for now.

sebhz commented:
The qzerty vs qwerty thing is because init_joystick_emulation() will modify the keyboard_normal table even if joystick emulation is disabled... mapping CPC_z to SDLK_z, and causing the issue.

This is easy to fix: we should only remap keys if joystick emulation is enabled and in this case we should take the mapping file into account to remap "whatever keysym happens to be bound to arrows, x and z" rather than hardcoding SDLK_x and SDLK_w. I'll write a patch for this tonight.
